samples/subsys/usb/hid-cdc/src/main.c

```diff
@@ -0,0 +1,868 @@
+{
+	chr_ptr_mouse = 0;
+	memset(data_buf_mouse, 0, sizeof(data_buf_mouse));
+}
+
+static void flush_buffer_kbd(void)
+{
+	chr_ptr_kbd = 0;
+	memset(data_buf_kbd, 0, sizeof(data_buf_kbd));
+}
+
+static void write_data(struct device *dev, const char *buf, int len)
+{
+	uart_irq_tx_enable(dev);
+
+	while (len) {
+		int written;
+
+		data_transmitted = false;
+		written = uart_fifo_fill(dev, (const u8_t *)buf, len);
+		while (data_transmitted == false) {
+			k_yield();
+		}
+
+		len -= written;
+		buf += written;
+	}
+
+	uart_irq_tx_disable(dev);
+}
+
+static void cdc_mouse_int_handler(struct device *dev)
+{
+	uart_irq_update(dev);
+
+	if (uart_irq_tx_ready(dev)) {
+		data_transmitted = true;
+	}
+
+	if (uart_irq_rx_ready(dev)) {
+		u32_t bytes_read;
+
+		while ((bytes_read = uart_fifo_read(dev,
+		       (u8_t *)data_buf_mouse+chr_ptr_mouse,
+		       sizeof(data_buf_mouse)-chr_ptr_mouse))) {
+			chr_ptr_mouse += bytes_read;
+			if (data_buf_mouse[chr_ptr_mouse-1] == 0x0D) {
+				/* ENTER */
+				struct app_evt_t *ev = app_evt_alloc();
+
+				data_buf_mouse[chr_ptr_mouse-1] = '\0';
+
+				if (!strcmp(data_buf_mouse, "up")) {
+					ev->event_type = CDC_UP;
+				} else if (!strcmp(data_buf_mouse, "down")) {
+					ev->event_type = CDC_DOWN;
+				} else if (!strcmp(data_buf_mouse, "right")) {
+					ev->event_type = CDC_RIGHT;
+				} else if (!strcmp(data_buf_mouse, "left")) {
+					ev->event_type = CDC_LEFT;
+				} else {
+					ev->event_type = CDC_UNKNOWN;
+				}
+				flush_buffer_mouse();
+				app_evt_put(ev);
+				k_sem_give(&evt_sem);
+			}
+
+			if (chr_ptr_mouse >= sizeof(data_buf_mouse)) {
+				LOG_WRN("Buffer overflow");
+				flush_buffer_mouse();
+			}
+		}
+	}
+}
+
+static void cdc_kbd_int_handler(struct device *dev)
+{
+	uart_irq_update(dev);
+
+	if (uart_irq_tx_ready(dev)) {
+		data_transmitted = true;
+	}
+
+	if (uart_irq_rx_ready(dev)) {
+		u32_t bytes_read;
+
+		while ((bytes_read = uart_fifo_read(dev,
+		       (u8_t *)data_buf_kbd+chr_ptr_kbd,
+		       sizeof(data_buf_kbd)-chr_ptr_kbd))) {
+			chr_ptr_kbd += bytes_read;
+			if (data_buf_kbd[chr_ptr_kbd-1] == 0x0D) {
+				/* ENTER */
+				struct app_evt_t *ev = app_evt_alloc();
+
+				data_buf_kbd[chr_ptr_kbd-1] = '\0';
+				strcpy(string, data_buf_kbd);
+				ev->event_type = CDC_STRING;
+				flush_buffer_kbd();
+				app_evt_put(ev);
+				k_sem_give(&evt_sem);
+			}
+		}
+	}
+}
+
+/* Devices */
+
+static void left_button(struct device *gpio, struct gpio_callback *cb,
+			u32_t pins)
+{
+	struct app_evt_t *ev = app_evt_alloc();
+
+	ev->event_type = GPIO_BUTTON_0,
+	app_evt_put(ev);
+	k_sem_give(&evt_sem);
+}
+
+#ifdef SW1_GPIO_PIN
+static void right_button(struct device *gpio, struct gpio_callback *cb,
+			 u32_t pins)
+{
+	struct app_evt_t *ev = app_evt_alloc();
+
+	ev->event_type = GPIO_BUTTON_1,
+	app_evt_put(ev);
+	k_sem_give(&evt_sem);
+}
+#endif
+
+#ifdef SW2_GPIO_PIN
+static void x_move(struct device *gpio, struct gpio_callback *cb, u32_t pins)
```

Comment:
All these functions are copy-pasta. Can we use parameter instead?

Reply:
"Copy-pasta", yum!